### ISSUE TITLE
Delay in Extension Programming

### ISSUE THREAD

**Issue by sweeneyapps** [rainforestapp/tester-chrome-extension#315]
I noticed when jobs were actually FOUND, the extension's button turned "YES" (green), there's delay on opening a new TAB.    So I looked into your source code and discovered that there's no delay between those actions.   Because I wrote my own extension to start with... I discovered that when I request a new tab, it opens ASAP!.   I tried another test to see if it's possible to delay the tab from opening asap...  I found none of the possibility to delay it.    

The delay is really bad...  I count in my head to 10 secs at most..    

Also sometimes, I leave tab open after a job,  the refresh part of the programming is really bad!   It doesn't refresh the current tab.   

I'll paste your code below here and you'll see what I'm looking at.    The delay shouldn't happen under this 2 lines of codes.

```javascript
if (resp.work_available) {
      chrome.browserAction.setBadgeBackgroundColor({color: GREEN});
      chrome.browserAction.setBadgeText({text: 'YES'});
      app.openOrFocusTab(resp.url);
    }
```

You can see when checking for job, there's no delay between browserAction.setBadgeText() and app.openOrFocusTab()

Let look into app.openOrFocusTab() below: 

```javascript
function openOrFocusTab(url) {
  if (appState.workTab === null) {
    app.makeNewWorkTab(url);
  } else {
    app.refreshTabInfo();
  }
}
```

again, no delay there..   let look into app.makeNewWorkTab() below:

```javascript
function makeNewWorkTab(url) {
  // make a new tab
    chrome.tabs.create({url}, t => {
    appState.workTab = t;
  });
}
```

Finally, we re at chrome.tabs.create(),  there's no delay into creating a new tab neither.

My gut feeling, your code is bloated with all other codes running in the background.   I would suggest to split the codes into 2 or 3 background tasks.

**Comment by helloworld111gh** [comment]
extension is open source, if you can make a better extension, then make one & do PR.
personally I do not have a lot of lag from "YES" to open tab.

**Comment by qpPykT** [comment]
And what about the delay between when a new tab is already open and receive a response with the hit or with any message. A small circle on the right side at the top, it can be spinning about 10-20 seconds. Is it same problem or not?

**Comment by sweeneyapps** [comment]
Alex, this is different problem.

spinning wheel loading is server side problem...  Nothing to do with Google Chrome Extension.     This extension only connect with the server to supply user information such as idle time, and ping the server for job availability.     It's like the extension asking the server, You got job?  Server just response back either Yes or No.   If Yes, then the extension will open new tab or refresh the tab w/ url of the job.

Your problem is server-side..   nothing to do w/ extension at all.   Because you are already receiving new URL and your tab is already loaded with Rainforest's UI client which is waiting to load more information from the server.

**Comment by sweeneyapps** [comment]
Richard,  I'll love to improve this extension but I don't work for Rainforest and it's their job to improve their own extension.    Yes, It's open source w/ poor documentation or closed documentation.   Also there's a lot of area where they could have done better other than the extension itself.   Such as how they deal with virtual machine availability and job ownership.    We should not see the "Virtual Machine not available" message AT ALL!    Just because if Rainforest is running a job, they should make sure all Virtual Machine are accounted for.   There's another work-around for it, like they could just assign the job to tester who already loaded the job and let them wait for the next available virtual machine.   That will make all of us happier.   If we don't want to wait, that's ok,  we can click to leave the job to other tester or if we idle on that job, we'll lose it.

**Comment by GreekGuy** [comment]
app.openOrFocusTab(resp.url);  You wait for the server to send the url so there is delay like propagation delay.

**Comment by tulko** [comment]
I think it has to do with each particular computer. I made changes to my computer (configurations) greatly improved the overall performance of the virtual machine, also in my router and I'm not connected directly to the Internet.
As for the delay in loading the site. I think it is because sometimes have too much traffic and saturates.
Regarding the case when set to YES and I will not have a tab, it is also a server problem and not the extention. Since you need the correct codes to access the test and this "url" send you the server.
I think it works like this:
1. terminal ping "RF" asks if there is a job
2. server rf answer "Yes"
3.The result extencion changes to YES
4. extencion expects RF will send the address (there is the delay in opening a new tab)
5. extencion receives the address with the number of assignment
6. Chrome attempts to load the site

So, sometimes you run out of the test because there is a delay in delivering the url so you can load the test. and the time delay chrome to ask the terminal. In all those seconds can someone comes before you.
you may get the url, chrome try asking the test when it does there is no longer any work

All problems are on the server.

I recommended them to use channels, assigning more jobs who have better punctuation. For example: seniority, number of test, rejections, bugs, etc.

If you had only three channels, for example

Channel 1: testers with high punctuation, which will be fewer
Channel 2: Intermediate punctuation
Channel 3: Low punctuation

If there are 1000 test to run at a certain time it is divided into 3

Channel 1: 334 test As this channel there are fewer testers, they receive more jobs
Channel 2; 333 testers test some more for the same amount of test
Channel 3: 333 test many more testers, for the same number of test

Calculated as the punctuation? are engineers have to take a statistical

For example:

Channel 1: 50000 or more points
Channel 2: 30000-50000
Channel 3: less than 30000

a possible calculation:
1. If I have 2087 test
2 If I have 235 bugs found
3 If I'm tester 21 months ago
4 If I have more than 2000 tests performed, or more than 1000 or less than one thousand (in this case over 2000)
5. If I have 3 rejections

As a result we have:
2087 + 235 = 2322
2322 x 2 (for point 4 to remove him 0) = 4644
4644 x 21 = 97524
97524/3 = 32508


Channel 2: 32508 points

If I get just one more, rejection would drop to channel 3 until the expiry of a rejection.

Having very least testers on channel 1, if there are more hits than testers at some point, the remaining test pass the channels immediately below.

Also can be delivered points for collaboration by contributing ideas, customers can qualify for any particular test, etc.

Anyway, I think they already have such a system, but do not say

"There are no virtual machines currently are available. Please retry in a few minutes." It is a fancy way of saying, "This test is not for you"

Why do I say this? It is that sometimes there are very few test in execution but still get the same message.
If no virtual machines should be assigned the test but you should expect to have some virtual machine.

I hope you understand, it's just a suggestion.
This comment is only a spirit of collaboration with RF not mean to criticize or harm.
I hope understood something, my English is pretty bad.

**Comment by nerdatwork** [comment]
@tulko I am humbled by the amount of energy and thoughts you have put into this idea. To feed your curiousity Rainforest is already working on a system called "Schrute". This system will give priority to testers who are online for most of the time among other things thereby eliminating testers complaining of sitting for hours and not getting any job. I hope that was helpful.

@sweeneyapps I thank you on behalf of fellow testers in taking time to analyze and giving your valuable input into how to improve the working of the extension.

**Comment by sweeneyapps** [comment]
@tulko you are incorrect on the part where extension had to wait until RF send a URL.   I completely test RF extension.   The extension check for the job, and the server ping back Yes or No, if Yes, it does include the job url, then the badge will turn Green.  There is supposed to be NO DELAY on opening a tab w/ the url supplied even if the server is still loading.   I'll include my own job checker extension which proven my theory correctly.

@nerdatwork  your welcome..  I'm including my codes below...  yup I took my time because I'm a javascript developer.

link to my github page w/ my own code proving that there's no delay between toggling badge on and creating a new tab w/ nodejs server mocking delay in page load.   The tab will open ASAP and code is just the same w/ RF's extension code without their bloated codes.   
https://github.com/sweeneyapps/JobCheckerForRainForest

**Comment by nberchev** [comment]
@tulko Can I ask you what changes you made to ur computer (configurations) which greatly improved the overall performance of the virtual machine?

**Comment by tulko** [comment]
@sweeneyapps I'm not desarrallor, although it is something I would like to learn. What I tell is by simple observation and comments gitub.
If YES and extention was not open any tab, because another reason to be ?.  yourself have said that in the code there is no delay. There is another reason, I do not want to mention why I believe this to YES if not the tab is opened is because there no url to load. Or maybe you could tell me if the numbers of "assignment" are generated the extention?

@nerdatwork   forgives smoke something last night and was inspired.  about "Schrute"  russ mention it in June AMA.  Surely this developing. I'm a fan RF. I'm always thinking of things to improve. And I always want to have more test daily. What I proposed priority to have better statistics. More old, Less rejections. and other issues. By having a channel system, can take into account many issues. can increase the number of points required from time to time. It is similar to the system of CF.

**Comment by helloworld111gh** [comment]
Just going to throw out a theory:
Maybe the delay is to ensure that when the tab opens you don't hit the rate
limit. Because if it opens as soon the yes is there, you might get rate
limit exceeded. So it waits a bit cool down that rate limiter.

**Comment by sweeneyapps** [comment]
@helloworld111gh   it's nothing to do with rate limit or anything.  Inside the code, whenever the Icon Button turn GREEN "YES", the only next step is to open a tab with a URL.   That's it.   Nothing to do with loading, server error or anything.   It is just something to do with javascript having hard time handling too much loads on all the scripts added into the memory.   Also it's handling too many events at once using one background thread.   That makes the extension programming slower to handle anything else.

I'll make it clear below..   I'll list steps in the codes in plain english.

1. Request Server for job information
2. Server response back saying yes or no, if yes: it response with plain text URL of the job.
3. Badge turn GREEN "YES" 
4. Chrome open new Tab
5. Chrome start loading URL.

This is what the code does.    Your theory on rate limit only will be shown on Step 5 with tab already opened.

**Comment by sweeneyapps** [comment]
I found some bugs :)   I will fix them and submit a pull request.

**Comment by sweeneyapps** [comment]
fixed some bugs! :) and did a pull request, https://github.com/rainforestapp/tester-chrome-extension/pull/339

all of testing is done on my "testing tab" extension on my github repo at https://github.com/sweeneyapps/TestWorkTab
